[10.x] `offsetExists` fetches attribute and not only checks

olivernybroe opened the pull request:
## Problem
When calling `offsetExists` on a model, it will call the underlying attribute and actually get the value. In the case where the getter is a bit more heavy this can result in unexpected extra load. `offsetGet` should be the method that actually fetches the attribute. 

```php
$model = new Model();

if(isset($model['name']) {
    return $model['name'];
}
```
The following code will result in two calls to the underlying attribute getter.  
This is mostly a problem when utilizing third party packages which accepts an `ArrayAccess` object and relies on `isset` logic. These packages now need custom logic for handling Laravel (for better performance), instead of just relying on the expected behaviour of `ArrayAccess`.

Commit:
test: isset on attribute also calls it

driesvints commented:
There seems to be a specific test for this situation which is now breaking which leads me to believe the current behavior is intended.

taylorotwell commented:
Please mark as ready for review when you have a code change. Thanks.

Commit:
fix: `offsetExists` checks if exist only

Change the logic of `offsetExists` so it checks if a property exists instead of actually fetching it also.

In the case of attribute getters we need to fetch the property still

olivernybroe commented:
@taylorotwell I have submitted a potential fix. 

My solution does include a breaking change as isset check on a model where the value of the attribute is `null` will now result in `true` instead of `false` (hence why I am targeting next major). 


This solution could be rewritten to not include a breaking change, but with the tradeoff that checking if an attribute exists on a model will still result in the mutator being called. However it would still fix the problem of doing `isset` on a relationship, so it won't actually run database queries on `isset` (unless you do it in inside your mutator).


I am proposing doing it with the breaking change as I think that is the most optimal solution, but if you would prefer the other solution, I can rewrite.

taylorotwell commented:
It seems like `isset` returning `true` on `null` is not desired? From the PHP docs on `isset`: "Determine if a variable is declared and **is different than null**."

spawnia commented:
I think the semantics of `isset()` pretty much force the existing implementation. It might just be impossible to optimize the code snippet you provided as an example.

However, I could see the addition of a method that contains the checks you put into `offsetExists()` here. Perhaps called `hasProperty()`?

olivernybroe commented:
@spawnia the improvements can still me made for relationships at least :) I will revert the attribute getters

spawnia commented:
So `isset()` for a relation that is defined but will return `null` when fetched will be `true`?

olivernybroe commented:
@spawnia you are right, I was only thinking about many to many relationships. but of course with belongs to relations it will be wrong. 

Closing this PR as my solution doesn't actually work.
